### PR TITLE
Attempt at providing an owned version of command

### LINKED ISSUE #117: Feature: A new `RemoteChildOwned` struct to avoid self-ref issue
From https://github.com/openssh-rust/openssh-sftp-client/pull/65#issuecomment-1533280291 :

> If it possible for Sftp to hold a session and spawn command (which means hold on RemoteChild) internally? So that users don't need to hold session, remote_child and Sftp at the same time. To avoid RemoteChild been dropped earlier than Sftp, users like opendal have to use owning_ref tricks.

`openssh_sftp_client::Sftp` now has a new method `new_with_auxiliary` to hold any data and @Xuanwo & @silver-ymz is currently using a [workaround I proposed](https://github.com/openssh-rust/openssh-sftp-client/pull/65#issuecomment-1535137285) by using a `Box`ed future.

But I think it's still better to have a `RemoteChildOwned` to avoid the self-ref issue, it's pretty bad for anybody using it with `openssh_sftp_client`.

Alternatively, we can wait for the new proxy-mode to openssh-mux-client https://github.com/openssh-rust/openssh-mux-client/issues/6 , which will enable you to use stdin/stdout/stderr while dropping Session and RemoteChild.

It will also support windows.

Unfortunately, it was blocking on `tokio_util::sync::WaitForCancellationFutureOwned`, so I switched to other projects and I kind of forgot it.
I will restart working on it, but it will takes quite some time before it's ready to use.

### PR DESCRIPTION
**Pull request by masklinn**
Pushing this to run the tests as getting things to work locally on a mac seems like a bit of a bear.

- Could not get the lifetimes right to not break `RemoteChild::session` (the lifetime becomes bound to the child's rather than the original session's), so move it into its own impl block, and added an extremely debatable `Child::into_session`, since it means not being able to call `wait`, or `wait_with_output`, or `disconnect`, so probably not useful (but a slightly separate variant of `Child::session` seems awful).
- Initially wanted to have a version directly owning a `Session`, but that turns out impossible as the `Command` builder works off of mutable references, so not possible to consume the builder in order to get the session out, short of creating either separate `impl` blocks for `Command<&Session>`, `Command<Arc<Session>>` and `Command<Session>`, or separate consuming spawns, neither sounding great.
- Naming is dubious, especially `shared_command`, but I don't recall ever seeing a method around shared refs so I went simple. I would have impl'd `Arc<Session>::into_command` but that's not legal...

  Maybe a free function would be better?

Hopefully will eventually fix #117

### CODE REVIEW

**Comment by jonhoo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/openssh-rust/openssh/136)

**Comment by masklinn** [comment]
@NobodyXu can you approve workflow runs? Also maybe comment on this early starting point.

**Comment by NobodyXu** [comment]
P.S. I don't think implementing `Command<Session>` is a good idea, since it's never designed to be taken by value.

`Command<Arc<Session>>` or `Command<Rc<Session>>` is more similar to `Command<&Session>` and I'd like to see support for them.

File: src/child.rs
Comment (NobodyXu) [inline src/child.rs:88]:
```suggestion
impl<S: AsRef<Session> + Clone> Child<S> {
```

File: src/child.rs
Comment (NobodyXu) [inline src/child.rs:78]:
```suggestion
pub struct Child<S> {
    session: S,
```

File: src/child.rs
Comment (NobodyXu) [reply src/child.rs:88]: With the `Clone` bound, you can also implement `fn session() -> S`.

File: src/session.rs
Comment (NobodyXu) [inline src/session.rs:280]: I think `shared_thread_safe_command` is a better name.
Then we can have `shared_command` for `Rc<...>`

File: src/session.rs
Comment (NobodyXu) [inline src/session.rs:284]: I would like to also have another generic version of this that takes `this: impl AsRef<Session> + Clone`

File: tests/openssh.rs
Comment (NobodyXu) [inline tests/openssh.rs:1064]:
```suggestion
async fn arc_client() {
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/child.rs
Comment (masklinn) [reply src/child.rs:88]: Yeah I figured that afterwards but didn't have the time to update the PR.

I don't think the `AsRef` bound is useful since at the end fo the day the `Client` doesn't actually care about the `Session`, beyond the outstanding reference in the borrow case. And the `Clone` case is only useful for `Session` itself.

I guess I could bound the entire thing anyway to ensure `Child` is only used with a session-adjacent object, but it feels unnecessary?

File: src/session.rs
Comment (masklinn) [reply src/session.rs:284]: I'm not sure `impl` can be used this way, and I'm reasonably certain it it's not supported (yet) for a `self`. It could be a free or factory function, but in that case maybe it should be a public constructor for `OwnedCommand` instead?

**Comment by masklinn** [comment]
> P.S. I don't think implementing `Command<Session>` is a good idea, since it's never designed to be taken by value.

I was mostly thinking of the case where a user might want to run a single command on the session, but didn't want to bother with borrowing. However as I wrote up it doesn't work without changing the `Command` API entirely so that's moot.

File: src/session.rs
Comment (masklinn) [reply src/session.rs:284]: [Currently valid list of self types](https://doc.rust-lang.org/reference/items/associated-items.html#methods) is `Self`, reference, box, rc, arc, and pin. So yeah no impl as an instance method.

File: src/session.rs
Comment (masklinn) [reply src/session.rs]: Added that change but TBH it feels unwieldingly verbose.

File: src/child.rs
Comment (NobodyXu) [reply src/child.rs]: That's true, so it's probably ok for `Child` to not constraint `S` since it is not used.

File: src/session.rs
Comment (NobodyXu) [reply src/session.rs]: You can make it an associated function instead `this: impl AsRef<Session> + Clone` is valid.

Comment (masklinn) [reply src/session.rs]: Oh I see that's what you meant, sorry I misunderstood.

Should it just replace the Arc version then? As it kinda obviates it, and if we call it "to_command" or something like that, even with an explicit `Arc` wrapper it's a lot shorter still.

Comment (NobodyXu) [reply src/session.rs]: You could keep it as a convenient method, or remove it if you think it's unnecessary.

File: src/session.rs
Comment (NobodyXu) [reply src/session.rs]: Hmmm, maybe arc_command is a better name?

File: src/session.rs
Comment (masklinn) [reply src/session.rs]: I'll be honest here, I don't use the crate, I just saw the issue during jonhoo's stream and figured I had an idea which could solve the problem with not too much code duplication (or macro).

So while I do have my own sensibilities about code and APIs, I'd *much* rather go with what you, jon, and other actual users of the crate (maybe the other maintainers of openssh-sftp-client since that's where things started?) would prefer.

Comment (NobodyXu) [reply src/session.rs]: In that case, let's keep the arc_command, AFAIK many do use it from multi-threaded async context with Arc.

Comment (masklinn) [reply src/session.rs]: OK if it's a common use case it makes sense to have a convenient shortcut. And the associated function does look pretty good in the test.

I made the arc versions just be convenience shortcuts over the associated function as that seems simpler / clearer.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:210]: Can we rename this to `CommandAnyClonable`?

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/command.rs
Comment (masklinn) [reply src/command.rs]: Ah sorry I had not seen your comment before doing a bit more shuffling around, squashing the commits, and writing docs.

I ended up renaming the internal command back to `Command`, and doing the aliasing on the edge in `lib.rs`, with the inner generic `Command` being exported as `OwningCommand` (seems better than `Owned`), and the `openssh::RemoteChild` and `openssh::Command` aliases being created there (`lib.rs`) rather than inside the crate guts.

I can still rename `crate::command::Command` if you prefer some other name tho.

File: src/session.rs
Comment (NobodyXu) [inline src/session.rs:275]: You can forward command and raw_command to to_command and to_raw_command

File: src/session.rs
Comment (NobodyXu) [inline src/session.rs:404]: Can we also have a generic version for subsystem please?

It cannot be emulated by to_command or to_raw_command

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:187]: Did you forget to update the name?

File: src/lib.rs
Comment (NobodyXu) [inline src/lib.rs:170]: I suggest renaming the type instead of creating a type alias, it would make our doc confusing to user

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/command.rs
Comment (masklinn) [reply src/command.rs:187]: Yep, looks like I missed this one.

File: src/session.rs
Comment (masklinn) [reply src/session.rs:275]: Interestingly no, because there is no `AsRef<T> for &T`.

File: src/session.rs
Comment (masklinn) [reply src/session.rs:404]: Sure, should it have both the `to_` and the `arc` version, or just the `to_` version?

File: src/lib.rs
Comment (masklinn) [reply src/lib.rs:170]: Damn and here I'd thought I could avoid renaming `Command` everywhere.

File: src/session.rs
Comment (NobodyXu) [reply src/session.rs:404]: Just to to version will be good enough.

File: src/session.rs
Comment (NobodyXu) [reply src/session.rs:275]: Hmmm, that's unexpected

Comment (masklinn) [reply src/session.rs:275]: Yep it also surprised me when I hit it. One of the reasons I'm always a little bit confused about using AsRef v Borrow v Deref.

Here AsRef might be the wrong choice because it allows "transverse" ref conversions e.g. `AsRef<OsStr> for str` and has more flexibility but I'm not sure that's needed.

`Borrow` and `Deref` are more limited but might be more suitable.

Comment (NobodyXu) [reply src/session.rs:275]: In that case, maybe swap that to Deref, since reference/Arc/Rc etc all have deref implemented.

**Comment by masklinn** [comment]
- moved the trait bound from AsRef to Deref
  - made `[raw_]command` go through `to_[raw_]command` since they now can
- renamed the owning `Command` to `OwningCommand` instead of just aliasing it as the crate root, for documentary clarity
- added a `to_subcommand` associated function to do the session-owning thing for subcommand features

File: src/session.rs
Comment (NobodyXu) [inline src/session.rs:307]: I think taking a 

```suggestion
        self: &std::sync::Arc<Session>,
```

is better and more ergonomic, otherwise users would have to do `session.clone().arc_command(..)`

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
BTW, you need to fix the CI before I can merge this PR.

File: src/session.rs
Comment (masklinn) [reply src/session.rs:307]: I think the current version is better, because otherwise the arc has to be cloned internally, so if the caller just wants to create a movable command/child they can't just move the one arc they have they get additional and unnecessary refcount traffic.

**Comment by masklinn** [comment]
> BTW, you need to fix the CI before I can merge this PR.

Yep but since I can't run the tests locally I can only push and wait for CI approval. I'll get on that.

File: src/session.rs
Comment (NobodyXu) [reply src/session.rs:307]: That's reasonable.

**Comment by NobodyXu** [comment]
@masklinn There's a [`run_ci_tests.sh`](https://github.com/openssh-rust/openssh/blob/master/run_ci_tests.sh) in the repository that can be used to run some of the tests.

For linting, it's just `cargo-clippy` and `cargo-fmt`.

**Comment by masklinn** [comment]
> There's a [run_ci_tests.sh](https://github.com/openssh-rust/openssh/blob/master/run_ci_tests.sh) in the repository that can be used to run some of the tests.

Yeah but it requires docker, and I'm on a mac and not super willing to install docker desktop.

> For linting, it's just `cargo-clippy` and `cargo-fmt`.

Yeah I ran fmt but it changed imports I had not touched so I figured the repo had strange configuration and reverted them, and they show up as lint errors so I guess it's just that they were last fmt'd before a change in configuration or something.

**Comment by masklinn** [comment]
Hopefully CI is good now, I broke one of the existing doc comments by mistake, and the new `arc_command` example had a few brainfarts.

**Comment by NobodyXu** [comment]
> Yeah but it requires docker, and I'm on a mac and not super willing to install docker desktop.

You can use [OrbStack](https://orbstack.dev/), an alternative to docker desktop that claims to be faster than docker desktop and uses less energy.

I use it for docker desktop now and it indeed startup quickly.

> It's nbd as far as I'm concerned just a bit more latency but I'm not replying within seconds anyway. Unless you have to approve every CI run I can imagine that gets old.

Yes, I do need to approve every CI run.

Once this PR lands in main, any new PR from you will have CI run automatically without having me to approve.

> so I guess it's just that they were last fmt'd before a change in configuration or something.

It's probably because new lints are added to newer cargo.

File: tests/openssh.rs
Comment (NobodyXu) [inline tests/openssh.rs:1099]:
```suggestion
async fn rc_client() {
```

I think `rc_client` is a better name here.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
The coverage failure is probably due to CI setup (e.g. GHA does not allow PRs from non-member to use secrets) or whatever, I can bypass that and merge the PR once you address my review.

File: tests/openssh.rs
Comment (masklinn) [reply tests/openssh.rs:1099]: Dunno, the intent of the name was to indicate that we're testing the generic `Session::to_command` rather than the ref-specific Session::command` or the Arc-specific `Session::arc_command`, the Rc is an implementation detail of the test but I figured I might as well use that since references and arc are already exercised. Would it help if I made most of the body generic and then ran it with both Rc and Arc?

Comment (NobodyXu) [reply tests/openssh.rs:1099]: It makes sense then, though I would like to rename this and the other test fn  to test_session_to_command and test_session_arc_command

Comment (masklinn) [reply tests/openssh.rs:1099]: Works for me.

**Review by NobodyXu — APPROVED** [review]

**Review by NobodyXu — APPROVED** [review]

**Comment by NobodyXu** [comment]
Thank you @masklinn !

**Comment by NobodyXu** [comment]
Released in v0.10.1

**Pull request merged**
